### PR TITLE
[Feature]migrate xllm kernels

### PR DESCRIPTION
**Pull request by 2Sheldonlaw**
**Overview**
Compared 4 Ascend kernel implementations from the xllm repository against our repository, migrating the customer's superior implementations and test cases to tilelang-ascend.

**Changes**

**1. New causal_conv1d_decode operator (examples/causal_conv1d/causal_conv1d_decode.py)**
- Repository previously only had prefill version; decode version (single-step token processing) was missing
- Migrated from xllm, de-frameworked (removed register_kernel/DispatchField/generate_source), converted to JIT mode
- Includes 2 test cases (translated from xllm C++ gtest)

**2. fused_sigmoid_gating_delta_rule test case migration** (examples/fused_sigmoid_gating_delta_rule/fused_sigmoid_gating_delta_rule_varlen.py)
- Found that xllm uses fp32 for A_log/dt_bias/init_state (better precision) and T.tile.sigmoid primitive (more efficient). However, dtype changes would cause data flow inconsistency (requires full migration), so kernel was not modified for now
- Migrated 4 xllm C++ gtest test cases to varlen.py, expanding test coverage from 1 to 5 cases

**3. chunk_gated_delta_rule gating mask fix (examples_experiment/chunk_gated_delta_rule/expert_chunk_gated_delta_rule.py)**
- Fixed gating numerical bug: when g_diff > 0, original implementation directly computes exp(g_diff) which yields incorrect results (>1)
- Ported xllm's compare+select mask approach: when g_diff > 0, value is set to -inf, so exp(-inf)=0 gives correct result
- Updated golden reference to apply the same masking for consistency
- Added 2 new UB buffers (g_exp_ub_pad / g_mask_ub_pad) for compare alignment
- Migrated 7 xllm C++ gtest test cases, expanding test coverage from 1 to 8 cases

**4. Operators not migrated**
- causal_conv1d (prefill): Repository already has a superior implementation (double-buffered pipeline + 4-token batch processing); xllm version is a basic sequential implementation, no migration needed
- chunk_gated_delta_rule (overall): Architecture already consistent (Expert mode + Cube+Vector pipeline + double buffering); only gating mask differed, which has been fixed

**Test Verification**
All tests passed on NPU environment:
- causal_conv1d_decode.py: 2/2 PASS
- fused_sigmoid_gating_delta_rule_varlen.py: 5/5 PASS
- expert_chunk_gated_delta_rule.py: 8/8 PASS

### CODE REVIEW

File: examples_experiment/chunk_gated_delta_rule/expert_chunk_gated_delta_rule.py [hunk 1]
```diff
@@ -217,15 +220,21 @@ def main(
...
+                        T.tile.select(
+                            g_exp_ub_pad,
+                            g_mask_ub_pad,
+                            g_exp_ub_pad,
+                            -T.infinity(accum_dtype),
+                            "VSEL_TENSOR_SCALAR_MODE",
+                        )
+                        T.copy(g_exp_ub_pad[0 : BT // 2], g_exp_ub)
                         T.tile.exp(g_exp_ub, g_exp_ub)
-                        T.pipe_barrier("v")
```
Comment (bingmeiyou) [inline examples_experiment/chunk_gated_delta_rule/expert_chunk_gated_delta_rule.py:224]: 这里去掉之后, 可能导致PTO 模式下失败, 需要验证下PTO模式

Comment (2Sheldonlaw) [reply examples_experiment/chunk_gated_delta_rule/expert_chunk_gated_delta_rule.py:224]: 验证完成。加回 T.pipe_barrier("v") 后，PTO 模式也通过了

**Comment by 2Sheldonlaw** [comment]
### 对于[examples_experiment/chunk_gated_delta_rule/expert_chunk_gated_delta_rule.py]的修改说明：算子除了用例变了，kernel代码也变了，
**问题1：这样改完之后算子逻辑是不是变了？如果逻辑变了，是否需要写个新的算子进来？
问题2：为什么这样改？整合xllm的逻辑有没有最小变动方案？**

**1. 算子逻辑是否变了？是否需要新写算子？**
算子逻辑没变，是 bug 修复，不需要新写算子。gating 公式 exp(g_last - g) 没变，算法步骤（W@h → v_new=v-wh → gating → K^T@v_new → h+=）也没变。修改的只是 gating 计算中 g_diff > 0 时的边界处理：
  - g_diff <= 0（正常）：修改前后结果完全相同
  - g_diff > 0（异常）：修改前 exp(正数) > 1 会错误放大 v_new，修改后 compare+select 置为 -inf，exp(-inf)=0 清零
举例说明：
    gate = [0.002, -0.005, 0.003, 0.001]
    g = cumsum = [0.002, -0.003, 0.000, 0.001]
    g_last = 0.001
    g_diff = [-0.001, 0.004, 0.001, 0]

token 1: g_diff=0.004 > 0
    修复前: exp(0.004) = 1.004 → v_new 被放大 ❌
    修复后: exp(-inf) = 0 → v_new 清零 ✅
g_diff > 0 意味着该 token 的 gate 累积值比 chunk 末尾大，之后 gate 在下降，这个 token 不应被放大，应清零。xllm 的实现也有同样的掩码处理，说明这是算法本身的设计要求。
原测试用全负 gate 值（rand * -1.0），cumsum 单调递减，g_diff 永远 <= 0，碰巧不触发 bug。xllm 测试用正负混合 gate 值（randn * 0.002），才会触发 g_diff > 0。

**2. 为什么这样改？有没有最小变动方案？**
当前改动已是最小方案，只改了 3 处：
    - gating kernel：compare+select 掩码替换直接 exp（~10 行）
    - golden reference：torch.where + -inf 保持一致（~3 行）
    - 新增 2 个 UB buffer：g_exp_ub_pad / g_mask_ub_pad（2 行，compare 需要 256B 对齐）
其余全部不动（Cube+Vector 流水线、双缓冲、GEMM、状态更新、flag 同步、数据搬运）。如果只迁移测试用例不修复 bug，xllm 的测试数据会触发 g_diff > 0 导致测试失败，所以 bug 修复和测试用例迁移必须一起做。

**Review by fuhouyu-hw — APPROVED** [review]
lgtm

**Review by ChaoyangJi — APPROVED** [review]
approve
